Create 2354-streaming-file-transfer.md (#2354)
matrix-org/matrix-doc: proposals/2354-streaming-file-transfer.md

# MSC2354 - Device to device streaming file transfers

File transfer in matrix currently works by uploading a file to the server, which is then distributed over
the participating servers in the room, and is available through an HTTP get request to everyone who knows
the URL. By nescesity this has a maximum filesize and also results in a more or less permanent availablity
of said file. For larger files, and/or files that should only be sent point to point it may be desirable

> to be able to send from device to device. As matrix already implements WebRTC signalling for voip, this 

Reviewer: Can you add a link to https://matrix.org/docs/spec/client_server/r0.6.0#voice-over-ip for reference?

functionality can be replicated for streaming file transfers from device to device. It can even be possible
to send files not between devices of two matrix users, but two matrix devices owned by the same user. For 
this we can use the webrtc datachannel.

## Proposal

To allow the device to device streaming data tranfser we propose to add events mirroring setting up a
WebRTC VoIP session. 

> To start a file transfer a user first selects the file to be transferred, after which the sending client 

Reviewer: Not sure if this is supposed to be a separate paragraph or not.  Can you either add a blank line before this, or combine this line with the previous (and re-wrap)?

    "event_id": "$Rqnc-F-dvnEYJTyHq_iKxU2bZ1CI92-kuZq3a5lr5Zg",
    "origin_server_ts": 1432735824653,
    "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
    "sender": "@example:example.org",
    "type": "m.d2dfile.cancel",
    "unsigned": {
        "age": 1234
    }
}
```

## Potential issues
It's tempting to want to use this for device to device transfer in public rooms, 
but an evil homeserver could hijack the webrtc session by pretending to be a device 
for the intended recipient (if that user has an account on the evil server). As such this 
should be limited to private rooms.  
Otherwise the matrix logic is quite simple here, basically following the same flow as 1:1 VoIP calls,
however: most guides (including the synapse github how-to) advice to disable TCP relays in the turnserver,
which may be unwanted for file transfers.

While this proposal focusses on streaming file transfer, 
a webrtc datachannel could be used for any generic data transfer, 
for the use of any other application, such as games for instance. 
If such application is wanted, the d2dfile event type may be a poorly chosen name.
For this proposal other uses for a webrtc datachannel is deemed out of scope.

> 

Reviewer: Some more potential issues:

This method requires both sender and recipient to be online at the same time.

The recipient can only receive the file once, on one device.

Author: I would argue that recieving the file once on one device is basically the point of the method, but I added it anyway.

Note that I would explicitly like it if we can also use this as file transfer between devices owned by the same user, meaning that the sender and recipient are basically the same, except on different devices.

## Alternatives
The current way of sending files is a valid alternative, 
the biggest upside of adding this proposal is that this allows for 
streaming file transfer outside of matrix and is thus not limited in filesize 
and negates the need for the matrix server to host any of the files.  
Since filetransfer over matrix isn't unique with the device to device transfer,
and implementation of filetransfers over webrtc is probably not trivial, 
this feature could be marked as optional.

## Security considerations
Using RTCDataChannel to transfer files could be abused to send malware 
without having the possibility of checking for this in between on the serverside.

> This could be mitigated similarly to how dinsinc does this,

Reviewer: FYI, DINSIC has been renamed to DINUM

Author: And I typoed it in the first place. 😁

# MSC2354 - Device to device streaming file transfers

File transfer in matrix currently works by uploading a file to the server, which is then distributed over
the participating servers in the room, and is available through an HTTP get request to everyone who knows

> the URL. By nescesity this has a maximum filesize and also results in a more or less permanent availablity

Reviewer 2: ```suggestion
the URL. By necessity this has a maximum filesize and also results in a more or less permanent availability
```

> of said file. For larger files, and/or files that should only be sent point to point it may be desirable

Reviewer 2: ```suggestion
of said file. For larger files, and/or files that should only be sent from point-to-point, it may be desirable
```

to be able to send from device to device. As matrix already implements

> [WebRTC signalling for voip](https://matrix.org/docs/spec/client_server/r0.6.0#voice-over-ip), this 

Reviewer 3: ```suggestion
[WebRTC signalling for VoIP](https://matrix.org/docs/spec/client_server/r0.6.0#voice-over-ip), this 
```

functionality can be replicated for streaming file transfers from device to device. It can even be possible
to send files not between devices of two matrix users, but two matrix devices owned by the same user. For 

> this we can use the webrtc datachannel.

Reviewer 3: ```suggestion
this we can use the WebRTC datachannel.
```

## Proposal

To allow the device to device streaming data tranfser we propose to add events mirroring setting up a
WebRTC VoIP session.

To start a file transfer a user first selects the file to be transferred, after which the sending client 
sends an `m.d2dfile.invite` event with:

```
{
    "content": {
        "transfer_id": "12345",
        "lifetime": 60000,

>         "filename": "example.doc",

Reviewer 4: Would be nice to have a sha256sum field here, to nail down the exact content to be transferred.

Reviewer 3: I like the idea but think that it should be optional. I can imagine cases where the file is generated on the fly, for example by bots. Alternatively in embedded cases the resource usage may be undesirable. Even end-user-cases such as streaming a large video from network storage should not need to be blocked by downloading the first time to compute a hash and maybe requiring re-downloading some or all of the file to send it.

I also think it doesn't add much either since you are getting the offer on the same channel and the offer + WebRTC provide integrity. But having an additional checksum doesn't hurt.

Reviewer 5: Checksum checking should definitely be part of the transfer process, however it makes more sense to send any hashes *after* the transfer is complete.

        "info": {
            "mimetype": "application/msword",

>             "size": 46144

Reviewer 3: On a similar note, is size required? What if the transfer should be dynamically generated?

        },
        "offer": {
            "sdp": "v=0\r\no=- 6584580628695956864 2 IN IP4 127.0.0.1[...]",
            "type": "offer"
        },
        "version": 0
    },
    "event_id": "$Rqnc-F-dvnEYJTyHq_iKxU2bZ1CI92-kuZq3a5lr5Zg",
    "origin_server_ts": 1432735824653,
    "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
    "sender": "@example:example.org",

>     "type": "m.d2dfile.invite",

Reviewer 3: Nit: I'm not a huge fan of `d2dfile`. `d2d` isn't self-documenting, you would need to look up to find out what it means. How about `file_transfer`. I think "transfer" is a good name to emphasize that this is ephemeral and live. Other options are `live_file`, `direct_file`, `live_file_transfer` or `direct_file_transfer`.

    "unsigned": {
        "age": 1234
    }
}
```

This triggers other devices in the room to alert the user about the offered transfer, displaying the filename, filetype and filesize.

In order to establish an RTCDataChannel, the sending device can gather candidates. 
This is done following the spec detailing how to find a turnserver by sending an 
authenticated get request to `https://example.org/_matrix/client/r0/voip/turnServer`.
Following the same course as used for setting up a VoIP call, this returns 
the address of the turnserver and a username and password. This turnserver can then be used 
to find the peer to peer candidates and/or a relay. Otherwise the external (possibly local) 
ip address can be used to create a candidate.

These candidates are then sent using: 
`m.d2dfile.candidates` with:
```
{
    "content": {

>         "transfer_id": "12345",

Reviewer 6: Have you considered using a `m.reference` relation to refer to the original invite rather than a custom `transfer_id`? This is also how the recently added #2241 does it.

Author: I'd be fine either way, but the reason for using transfer_id is to mirror the voip webrtc spec.

Reviewer 6: I see. My concern was that clients will end up having to implement multiple ways of tracking relations. Given that this doesn't introduce a new way, it could be fine. I'm not sure but I'd assume changing `m.call.*` to use proper relations would not be feasible at this point, so I'm fine either way as well :)

Author: It could be worth changing `m.call.*`, but that's certainly out of the scope of this MSC. Thanks for the feedback!

        "transfer_id": "12345",
        "lifetime": 60000,
        "version": 0
    },
    "event_id": "$Rqnc-F-dvnEYJTyHq_iKxU2bZ1CI92-kuZq3a5lr5Zg",
    "origin_server_ts": 1432735824653,
    "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
    "sender": "@example:example.org",
    "type": "m.d2dfile.answer",
    "unsigned": {
        "age": 1234
    }
}
```
The accepting device also sends a `m.d2dfile.candidates` in order to establish the data connection.

To cancel a filetransfer or reject an offer the `m.d2dfile.cancel` event is sent, as follows:

```
{
    "content": {
        "transfer_id": "12345",
        "version": 0
    },
    "event_id": "$Rqnc-F-dvnEYJTyHq_iKxU2bZ1CI92-kuZq3a5lr5Zg",
    "origin_server_ts": 1432735824653,
    "room_id": "!jEsUZKDJdhlrceRyVU:example.org",
    "sender": "@example:example.org",
    "type": "m.d2dfile.cancel",
    "unsigned": {
        "age": 1234
    }
}
```

## Potential issues
It's tempting to want to use this for device to device transfer in public rooms, 
but an evil homeserver could hijack the webrtc session by pretending to be a device 
for the intended recipient (if that user has an account on the evil server). As such this 

> should be limited to private rooms.  

Reviewer 3: Should public/private be replaced with unencrypted/e2ee rooms? IIUC the main concern is that without e2ee the homeservers can send/accept transfers on behalf of their users.

Reviewer 7: i'm unclear on the risks associated here, but would the perceived risks be mitigated by offloading the actual file transfer to a webtorrent tracker endpoint, similarly to how TURN is required to negotiate calling?

Otherwise the matrix logic is quite simple here, basically following the same flow as 1:1 VoIP calls,
however: most guides (including the synapse github how-to) advice to disable TCP relays in the turnserver,
which may be unwanted for file transfers.

> Both sender and reciever need to be online simultaniously for this mode of file transfer
> to work, which is unexpected in the context of the existing matrix file transfer.

Reviewer 8: ```suggestion
Both sender and reciever need to be online simultaneously for this mode of file transfer
to work, which is unexpected in the context of the existing matrix file transfer.
```

> to work, which is unexpected in the context of the existing matrix file transfer.

Reviewer 5: Would it be possible to use the homeserver as intermediate storage? This would require less resources on the HS side than traditional file sending, while drastically relaxing any time constraints.

This file transfer will only send the file once, from one device to another.

While this proposal focusses on streaming file transfer, 

> a webrtc datachannel could be used for any generic data transfer, 

Reviewer 3: ```suggestion
a WebRTC datachannel could be used for any generic data transfer, 
```
